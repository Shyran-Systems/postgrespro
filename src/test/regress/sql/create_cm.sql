CREATE COMPRESSION METHOD jsonb2 HANDLER no_such_handler;

CREATE COMPRESSION METHOD jsonb HANDLER jsonb_handler;

CREATE COMPRESSION METHOD jsonb2 HANDLER jsonb_handler;

CREATE TABLE jstest(
	js text COMPRESSED jsonb
);

CREATE TABLE jstest(
	js text COMPRESSED jsonb2
);

CREATE TABLE jstest(
	js1 json,
	js2 json COMPRESSED jsonb2,
	js3 json COMPRESSED jsonbc,
	js4 json COMPRESSED jsonbc,
	js5 jsonb
);

INSERT INTO jstest
SELECT js, js, js, js, js
FROM (VALUES
	('{"key1": "val1", "key2": ["val2", 3, 4, 5]}'::json),
	('["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]'),
	('"val"'),
	('12345'),
	(('[' || repeat('"test", ', 10000) || '"test"]')::json)
) AS jsvals(js);

SELECT
	substring(js1::text for 100),
	substring(js2::text for 100),
	substring(js3::text for 100),
	substring(js4::text for 100),
	substring(js5::text for 100)
FROM jstest;

-- check page items size
CREATE EXTENSION pageinspect;

SELECT
	lp, lp_off, lp_flags, lp_len,
	CASE lp WHEN 5 THEN NULL ELSE t_data END
FROM
	heap_page_items(get_raw_page('jstest', 0));

SELECT
	lp,
	length(unnest(t_attrs)),
	CASE lp WHEN 5 THEN NULL ELSE unnest(t_attrs) END
FROM
	heap_page_item_attrs(
		get_raw_page('jstest', 0),
		'jstest'::regclass,
		false
	);

DROP EXTENSION pageinspect;

-- copy json values with different compression

INSERT INTO jstest SELECT js1, js2, js3, js4, js5 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js2, js3, js4, js5, js1 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js3, js4, js5, js1, js2 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js4, js5, js1, js2, js3 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js5, js1, js2, js3, js4 FROM jstest LIMIT 5;

SELECT
	substring(js1::text for 100),
	substring(js2::text for 100),
	substring(js3::text for 100),
	substring(js4::text for 100),
	substring(js5::text for 100)
FROM jstest;



-- Try to drop compression method: fail because of dependent objects
DROP COMPRESSION METHOD jsonb2;

-- Drop compression method cascade
DROP COMPRESSION METHOD jsonb2 CASCADE;

SELECT * FROM jstest LIMIT 0;

DROP TABLE jstest;

-- Test ALTER TYPE SET COMPRESSED
ALTER TYPE json SET COMPRESSED jsonb;
CREATE TABLE jstest (js json);

SELECT attcompression FROM pg_attribute WHERE attrelid = 'jstest'::regclass AND attnum = 1;

INSERT INTO jstest VALUES ('[ 123,  "abc", { "k" : "v" }  ]');
SELECT * FROM jstest;
DROP TABLE jstest;

ALTER TYPE json SET NOT COMPRESSED;
CREATE TABLE jstest (js json);

SELECT attcompression FROM pg_attribute WHERE attrelid = 'jstest'::regclass AND attnum = 1;

INSERT INTO jstest VALUES ('[ 123,  "abc", { "k" : "v" }  ]');
SELECT * FROM jstest;
DROP TABLE jstest;

-- Test compressable type creation
CREATE TYPE json2;

CREATE TEMP TABLE json2_procs AS
SELECT * FROM pg_proc p WHERE proname IN ('json_in', 'json_out', 'json_null_cm_handler');

UPDATE json2_procs
SET proname = replace(proname, 'json_', 'json2_');

UPDATE json2_procs
SET prorettype = (SELECT oid FROM pg_type WHERE typname = 'json2')
WHERE proname = 'json2_in';

UPDATE json2_procs
SET proargtypes = (SELECT oid::text::oidvector FROM pg_type WHERE typname = 'json2')
WHERE proname = 'json2_out';

INSERT INTO pg_proc
SELECT * FROM json2_procs;

CREATE COMPRESSION METHOD json2_null HANDLER json2_null_cm_handler;

CREATE TYPE json2 (
	INPUT  = json2_in,
	OUTPUT = json2_out,
	NULLCM = json2_null
);

CREATE TEMP TABLE tjson2(js json2);
INSERT INTO tjson2 VALUES ('abc');
INSERT INTO tjson2 VALUES ('["abc", {"key": 123}, null]');
SELECT * FROM tjson2;

DROP FUNCTION json2_null_cm_handler(internal);
DROP FUNCTION json2_in(cstring);
DROP FUNCTION json2_out(json2);
DROP FUNCTION json2_out(json2) CASCADE;
DROP FUNCTION json2_in(cstring);
DROP FUNCTION json2_null_cm_handler(internal);
DROP FUNCTION json2_null_cm_handler(internal) CASCADE;

DROP TABLE tjson2;

-- Test compression methods on domains
CREATE DOMAIN json_not_null AS json NOT NULL;

CREATE TEMP TABLE json_domain_test1(js json_not_null);
SELECT attcompression FROM pg_attribute WHERE attrelid = 'json_domain_test1'::regclass AND attnum = 1;
DROP TABLE json_domain_test1;

CREATE TEMP TABLE json_domain_test2(js json_not_null compressed jsonb);
SELECT attcompression FROM pg_attribute WHERE attrelid = 'json_domain_test2'::regclass AND attnum = 1;
DROP TABLE json_domain_test2;

-- Test compression inheritance
create temp table json_parent1(js json);
create temp table json_parent2(js json compressed jsonb);
create temp table json_parent3(js json compressed jsonb);

create temp table json_child1(js json) inherits (json_parent1);
create temp table json_child2(js json) inherits (json_parent2);
create temp table json_child3(js json compressed jsonb) inherits (json_parent1);
create temp table json_child4(js json compressed jsonb) inherits (json_parent2);
create temp table json_child5(js json) inherits (json_parent1, json_parent2);
create temp table json_child6(js json) inherits (json_parent2, json_parent3);
create temp table json_child7(js json compressed jsonb) inherits (json_parent2, json_parent3);

SELECT a.attrelid::regclass, a.attnum, c.cmname
FROM pg_attribute a JOIN pg_compression c ON a.attcompression = c.oid
WHERE a.attrelid::regclass::text LIKE 'json_child%' AND a.attnum = 1;

drop table json_parent1 cascade;
drop table json_parent2 cascade;
drop table json_parent3 cascade;

