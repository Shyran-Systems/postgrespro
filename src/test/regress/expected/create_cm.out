CREATE COMPRESSION METHOD jsonb2 HANDLER no_such_handler;
ERROR:  function no_such_handler(internal) does not exist
CREATE COMPRESSION METHOD jsonb HANDLER jsonb_handler;
ERROR:  compression method "jsonb" already exists
CREATE COMPRESSION METHOD jsonb2 HANDLER jsonb_handler;
CREATE TABLE jstest(
	js text COMPRESSED jsonb
);
ERROR:  jsonb compression method is only applicable to json type
CREATE TABLE jstest(
	js text COMPRESSED jsonb2
);
ERROR:  jsonb compression method is only applicable to json type
CREATE TABLE jstest(
	js1 json,
	js2 json COMPRESSED jsonb2,
	js3 json COMPRESSED jsonbc,
	js4 json COMPRESSED jsonbc,
	js5 jsonb
);
INSERT INTO jstest
SELECT js, js, js, js, js
FROM (VALUES
	('{"key1": "val1", "key2": ["val2", 3, 4, 5]}'::json),
	('["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]'),
	('"val"'),
	('12345'),
	(('[' || repeat('"test", ', 10000) || '"test"]')::json)
) AS jsvals(js);
SELECT
	substring(js1::text for 100),
	substring(js2::text for 100),
	substring(js3::text for 100),
	substring(js4::text for 100),
	substring(js5::text for 100)
FROM jstest;
                                              substring                                               |                                              substring                                               |                                              substring                                               |                                              substring                                               |                                              substring                                               
------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------
 {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}
 ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]
 "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"
 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345
 ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te
(5 rows)

-- check page items size
CREATE EXTENSION pageinspect;
SELECT
	lp, lp_off, lp_flags, lp_len,
	CASE lp WHEN 5 THEN NULL ELSE t_data END
FROM
	heap_page_items(get_raw_page('jstest', 0));
 lp | lp_off | lp_flags | lp_len |                                                                                                                                                                                                                                                                        t_data                                                                                                                                                                                                                                                                        
----+--------+----------+--------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  1 |   7912 |        1 |    274 | \x597b226b657931223a202276616c31222c20226b657932223a205b2276616c32222c20332c20342c20355d7da302000020040000800400000004000000300000506b6579316b65793276616c31040000400400008008000010080000100800001076616c322000000000800300200000000080040020000000008005002d110121016776616c3112210b0b0b76616c3206080a2d110121016776616c3112210b0b0b76616c3206080aa302000020040000800400000004000000300000506b6579316b65793276616c31040000400400008008000010080000100800001076616c32200000000080030020000000008004002000000000800500
  2 |   7616 |        1 |    289 | \x5f5b2276616c31222c20322c207b226b31223a20227631222c20226b32223a20327d2c2022222c20352c202236225dad0600004004000080080000102400005000000000080000100100000076616c312000000000800200020000200200008002000000020000000a0000106b316b327631000020000000008002002000000000800500362f1a210b47010b0976616c3104110311010b7631040a362f1a210b47010b0976616c3104110311010b7631040a36ad0600004004000080080000102400005000000000080000100100000076616c312000000000800200020000200200008002000000020000000a0000106b316b32763100002000000000800200200000000080050036
  3 |   7544 |        1 |     66 | \x0d2276616c2219010000500300008076616c0d041976616c0d041976616c19010000500300008076616c
  4 |   7464 |        1 |     80 | \x0d313233343527010000500a000090280000000180010029090d041bf2c0010d041bf2c00127010000500a00009028000000018001002909
  5 |   6424 |        1 |   1033 | 
(5 rows)

SELECT
	lp,
	length(unnest(t_attrs)),
	CASE lp WHEN 5 THEN NULL ELSE unnest(t_attrs) END
FROM
	heap_page_item_attrs(
		get_raw_page('jstest', 0),
		'jstest'::regclass,
		false
	);
 lp | length |                                                                                     unnest                                                                                     
----+--------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  1 |     44 | \x597b226b657931223a202276616c31222c20226b657932223a205b2276616c32222c20332c20342c20355d7d
  1 |     81 | \xa302000020040000800400000004000000300000506b6579316b65793276616c31040000400400008008000010080000100800001076616c32200000000080030020000000008004002000000000800500
  1 |     22 | \x2d110121016776616c3112210b0b0b76616c3206080a
  1 |     22 | \x2d110121016776616c3112210b0b0b76616c3206080a
  1 |     81 | \xa302000020040000800400000004000000300000506b6579316b65793276616c31040000400400008008000010080000100800001076616c32200000000080030020000000008004002000000000800500
  2 |     47 | \x5f5b2276616c31222c20322c207b226b31223a20227631222c20226b32223a20327d2c2022222c20352c202236225d
  2 |     86 | \xad0600004004000080080000102400005000000000080000100100000076616c312000000000800200020000200200008002000000020000000a0000106b316b32763100002000000000800200200000000080050036
  2 |     23 | \x2f1a210b47010b0976616c3104110311010b7631040a36
  2 |     23 | \x2f1a210b47010b0976616c3104110311010b7631040a36
  2 |     86 | \xad0600004004000080080000102400005000000000080000100100000076616c312000000000800200020000200200008002000000020000000a0000106b316b32763100002000000000800200200000000080050036
  3 |      6 | \x0d2276616c22
  3 |     12 | \x19010000500300008076616c
  3 |      6 | \x0d041976616c
  3 |      6 | \x0d041976616c
  3 |     12 | \x19010000500300008076616c
  4 |      6 | \x0d3132333435
  4 |     19 | \x27010000500a00009028000000018001002909
  4 |      6 | \x0d041bf2c001
  4 |      6 | \x0d041bf2c001
  4 |     19 | \x27010000500a00009028000000018001002909
  5 |    937 | 
  5 |     18 | 
  5 |     18 | 
  5 |     18 | 
  5 |     18 | 
(25 rows)

DROP EXTENSION pageinspect;
-- copy json values with different compression
INSERT INTO jstest SELECT js1, js2, js3, js4, js5 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js2, js3, js4, js5, js1 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js3, js4, js5, js1, js2 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js4, js5, js1, js2, js3 FROM jstest LIMIT 5;
INSERT INTO jstest SELECT js5, js1, js2, js3, js4 FROM jstest LIMIT 5;
SELECT
	substring(js1::text for 100),
	substring(js2::text for 100),
	substring(js3::text for 100),
	substring(js4::text for 100),
	substring(js5::text for 100)
FROM jstest;
                                              substring                                               |                                              substring                                               |                                              substring                                               |                                              substring                                               |                                              substring                                               
------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------
 {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}
 ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]
 "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"
 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345
 ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te
 {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}
 ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]
 "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"
 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345
 ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te
 {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}
 ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]
 "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"
 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345
 ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te
 {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}
 ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]
 "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"
 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345
 ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te
 {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}
 ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]
 "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"
 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345
 ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te
 {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}                                                          | {"key1": "val1", "key2": ["val2", 3, 4, 5]}
 ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]                                                       | ["val1", 2, {"k1": "v1", "k2": 2}, "", 5, "6"]
 "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"                                                                                                | "val"
 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345                                                                                                | 12345
 ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te | ["test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "te
(30 rows)

-- Try to drop compression method: fail because of dependent objects
DROP COMPRESSION METHOD jsonb2;
ERROR:  cannot drop compression method jsonb2 because other objects depend on it
DETAIL:  table jstest column js3 depends on compression method jsonb2
table jstest column js7 depends on compression method jsonb2
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
-- Drop compression method cascade
DROP COMPRESSION METHOD jsonb2 CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table jstest column js3
drop cascades to table jstest column js7
SELECT * FROM jstest LIMIT 0;
 js1 | js2 | js4 | js5 | js6 
-----+-----+-----+-----+-----
(0 rows)

DROP TABLE jstest;
SELECT relname, relkind FROM pg_class WHERE relname LIKE 'jstest%';
 relname | relkind 
---------+---------
(0 rows)

-- Test ALTER TYPE SET COMPRESSED
ALTER TYPE json SET COMPRESSED jsonb;
CREATE TABLE jstest (js json);
SELECT attcompression FROM pg_attribute WHERE attrelid = 'jstest'::regclass AND attnum = 1;
 attcompression 
----------------
           3403
(1 row)

INSERT INTO jstest VALUES ('[ 123,  "abc", { "k" : "v" }  ]');
SELECT * FROM jstest;
            js            
--------------------------
 [123, "abc", {"k": "v"}]
(1 row)

DROP TABLE jstest;
ALTER TYPE json SET NOT COMPRESSED;
CREATE TABLE jstest (js json);
SELECT attcompression FROM pg_attribute WHERE attrelid = 'jstest'::regclass AND attnum = 1;
 attcompression 
----------------
           3451
(1 row)

INSERT INTO jstest VALUES ('[ 123,  "abc", { "k" : "v" }  ]');
SELECT * FROM jstest;
               js                
---------------------------------
 [ 123,  "abc", { "k" : "v" }  ]
(1 row)

DROP TABLE jstest;
-- Test compressable type creation
CREATE TYPE json2;
CREATE TEMP TABLE json2_procs AS
SELECT * FROM pg_proc p WHERE proname IN ('json_in', 'json_out', 'json_null_cm_handler');
UPDATE json2_procs
SET proname = replace(proname, 'json_', 'json2_');
UPDATE json2_procs
SET prorettype = (SELECT oid FROM pg_type WHERE typname = 'json2')
WHERE proname = 'json2_in';
UPDATE json2_procs
SET proargtypes = (SELECT oid::text::oidvector FROM pg_type WHERE typname = 'json2')
WHERE proname = 'json2_out';
INSERT INTO pg_proc
SELECT * FROM json2_procs;
CREATE COMPRESSION METHOD json2_null HANDLER json2_null_cm_handler;
CREATE TYPE json2 (
	INPUT  = json2_in,
	OUTPUT = json2_out,
	NULLCM = json2_null
);
CREATE TEMP TABLE tjson2(js json2);
INSERT INTO tjson2 VALUES ('abc');
ERROR:  invalid input syntax for type json
LINE 1: INSERT INTO tjson2 VALUES ('abc');
                                   ^
DETAIL:  Token "abc" is invalid.
CONTEXT:  JSON data, line 1: abc
INSERT INTO tjson2 VALUES ('["abc", {"key": 123}, null]');
SELECT * FROM tjson2;
             js              
-----------------------------
 ["abc", {"key": 123}, null]
(1 row)

DROP FUNCTION json2_null_cm_handler(internal);
ERROR:  cannot drop function json2_null_cm_handler(internal) because other objects depend on it
DETAIL:  compression method json2_null depends on function json2_null_cm_handler(internal)
type json2 depends on compression method json2_null
table tjson2 column js depends on compression method json2_null
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP FUNCTION json2_in(cstring);
ERROR:  cannot drop function json2_in(cstring) because other objects depend on it
DETAIL:  type json2 depends on function json2_in(cstring)
table tjson2 column js depends on type json2
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP FUNCTION json2_out(json2);
ERROR:  cannot drop function json2_out(json2) because other objects depend on it
DETAIL:  type json2 depends on function json2_out(json2)
table tjson2 column js depends on type json2
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP FUNCTION json2_out(json2) CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to type json2
drop cascades to table tjson2 column js
DROP FUNCTION json2_in(cstring);
DROP FUNCTION json2_null_cm_handler(internal);
ERROR:  cannot drop function json2_null_cm_handler(internal) because other objects depend on it
DETAIL:  compression method json2_null depends on function json2_null_cm_handler(internal)
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP FUNCTION json2_null_cm_handler(internal) CASCADE;
NOTICE:  drop cascades to compression method json2_null
DROP TABLE tjson2;
-- Test compression methods on domains
CREATE DOMAIN json_not_null AS json NOT NULL;
CREATE TEMP TABLE json_domain_test1(js json_not_null);
SELECT attcompression FROM pg_attribute WHERE attrelid = 'json_domain_test1'::regclass AND attnum = 1;
 attcompression 
----------------
           3451
(1 row)

DROP TABLE json_domain_test1;
CREATE TEMP TABLE json_domain_test2(js json_not_null compressed jsonb);
SELECT attcompression FROM pg_attribute WHERE attrelid = 'json_domain_test2'::regclass AND attnum = 1;
 attcompression 
----------------
           3403
(1 row)

DROP TABLE json_domain_test2;
-- Test compression inheritance
create temp table json_parent1(js json);
create temp table json_parent2(js json compressed jsonb);
create temp table json_parent3(js json compressed jsonb);
create temp table json_child1(js json) inherits (json_parent1);
NOTICE:  merging column "js" with inherited definition
create temp table json_child2(js json) inherits (json_parent2);
NOTICE:  merging column "js" with inherited definition
create temp table json_child3(js json compressed jsonb) inherits (json_parent1);
NOTICE:  merging column "js" with inherited definition
ERROR:  column "js" has a compression method conflict
DETAIL:  json_null versus jsonb
create temp table json_child4(js json compressed jsonb) inherits (json_parent2);
NOTICE:  merging column "js" with inherited definition
create temp table json_child5(js json) inherits (json_parent1, json_parent2);
NOTICE:  merging multiple inherited definitions of column "js"
ERROR:  column "js" has a compression method conflict
DETAIL:  json_null versus jsonb
create temp table json_child6(js json) inherits (json_parent2, json_parent3);
NOTICE:  merging multiple inherited definitions of column "js"
NOTICE:  merging column "js" with inherited definition
create temp table json_child7(js json compressed jsonb) inherits (json_parent2, json_parent3);
NOTICE:  merging multiple inherited definitions of column "js"
NOTICE:  merging column "js" with inherited definition
SELECT a.attrelid::regclass, a.attnum, c.cmname
FROM pg_attribute a JOIN pg_compression c ON a.attcompression = c.oid
WHERE a.attrelid::regclass::text LIKE 'json_child%' AND a.attnum = 1;
  attrelid   | attnum |  cmname   
-------------+--------+-----------
 json_child1 |      1 | json_null
 json_child2 |      1 | jsonb
 json_child4 |      1 | jsonb
 json_child6 |      1 | jsonb
 json_child7 |      1 | jsonb
(5 rows)

drop table json_parent1 cascade;
NOTICE:  drop cascades to table json_child1
drop table json_parent2 cascade;
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table json_child2
drop cascades to table json_child4
drop cascades to table json_child6
drop cascades to table json_child7
drop table json_parent3 cascade;
